ocaml/xapi/repository_helpers.ml

```diff
@@ -1361,15 +1324,139 @@ let prune_updateinfo_for_livepatches livepatches updateinfo =
+
+let is_livepatch_failure = function
+  | Guidance.RebootHostOnLivePatchFailure
+  | Guidance.RebootHostOnXenLivePatchFailure
+  | Guidance.RebootHostOnKernelLivePatchFailure ->
+      true
+  | _ ->
+      false
+
+type pending_ops = {
+    host_get: unit -> Guidance.t list
+  ; host_add:
+         [ `reboot_host
+         | `reboot_host_on_livepatch_failure
+         | `restart_toolstack
+         | `reboot_host_on_xen_livepatch_failure
+         | `reboot_host_on_kernel_livepatch_failure ]
+      -> unit
+  ; host_remove:
+         [ `reboot_host
+         | `reboot_host_on_livepatch_failure
+         | `restart_toolstack
+         | `reboot_host_on_xen_livepatch_failure
+         | `reboot_host_on_kernel_livepatch_failure ]
+      -> unit
+  ; vms_get: unit -> (string * Guidance.t list) list
+  ; vm_add: string -> [`restart_device_model | `restart_vm] -> unit
+  ; vm_remove: string -> [`restart_device_model | `restart_vm] -> unit
+}
+
+let get_ops_of_pending ~__context ~host ~kind =
+  let get_pending_guidances_of_host ~get =
+    get ~__context ~self:host
+    |> List.map (fun g -> Guidance.of_pending_guidance g)
+  in
+  let get_pending_guidances_of_vms ~get =
+    Db.Host.get_resident_VMs ~__context ~self:host
+    |> List.map (fun self -> (self, Db.VM.get_record ~__context ~self))
+    |> List.filter_map (fun (ref, record) ->
+           match
+             ( record.API.vM_is_control_domain
+             , record.API.vM_power_state
+             , Helpers.has_qemu_currently ~__context ~self:ref
+             )
+           with
+           | false, `Running, true | false, `Paused, true ->
+               Some ref
+           | _ ->
+               None
+       )
+    |> List.map (fun vm_ref ->
+           let pending_guidances =
+             get ~__context ~self:vm_ref
+             |> List.map Guidance.of_pending_guidance
+           in
+           (Ref.string_of vm_ref, pending_guidances)
+       )
+  in
+  match kind with
+  | Guidance.Mandatory ->
+      let host_get () =
+        get_pending_guidances_of_host ~get:Db.Host.get_pending_guidances
+      in
+      let host_add value =
+        Db.Host.add_pending_guidances ~__context ~self:host ~value
+      in
+      let host_remove value =
+        Db.Host.remove_pending_guidances ~__context ~self:host ~value
+      in
+      let vms_get () =
+        get_pending_guidances_of_vms ~get:Db.VM.get_pending_guidances
+      in
+      let vm_add vm value =
+        Db.VM.add_pending_guidances ~__context ~self:(Ref.of_string vm) ~value
+      in
+      let vm_remove vm value =
+        Db.VM.remove_pending_guidances ~__context ~self:(Ref.of_string vm)
+          ~value
+      in
+      {host_get; host_add; host_remove; vms_get; vm_add; vm_remove}
+  | _ ->
+      raise Api_errors.(Server_error (internal_error, ["Not implemented kind"]))
+
+let set_pending_guidances ~ops ~coming =
+  let pending_of_host =
+    ops.host_get () |> List.filter (fun x -> not (is_livepatch_failure x))
+  in
```

Comment:
I think any livepatch guidances in `host.pending_guidances` have been "updated" in the first [commit](https://github.com/xapi-project/xen-api/pull/5276/files#diff-0b5dc72783f904599eeed5fdcd93fbdcf368b71edc18f657d123a8982bc5402c) of the feature branch?

Reply:
Yes. But this is a general logic for all kinds of guidance.